x/virtualgroup/keeper/grpc_query.go

```diff
@@ -119,3 +119,52 @@ func (k Keeper) AvailableGlobalVirtualGroupFamilies(goCtx context.Context, req *
 	}
 	return &types.AvailableGlobalVirtualGroupFamiliesResponse{GlobalVirtualGroupFamilyIds: availableFamilyIds}, nil
 }
+
+func (k Keeper) SwapInInfo(goCtx context.Context, req *types.QuerySwapInInfoRequest) (*types.QuerySwapInInfoResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	swapInInfo, found := k.GetSwapInInfo(ctx, req.GetGlobalVirtualGroupFamilyId(), req.GetGlobalVirtualGroupId())
+	if !found {
+		return nil, types.ErrSwapInInfoNotExist
+	}
+	if uint64(ctx.BlockTime().Unix()) >= swapInInfo.ExpirationTime {
```

Comment:
Can we also prune the expired swap in infos?

Reply:
Dont need extra process to do the pruning. The expired one will be override by the new successor's swapIn info. And also keep it here in case the same SP re-apply and block the exit process.